consensus/src/quorum_store/utils.rs

```diff
@@ -341,50 +373,55 @@ impl ProofQueue {
         );
     }
 
-    pub(crate) fn add_batch_summaries(
+    pub fn insert_batches(
         &mut self,
-        batch_summaries: Vec<(BatchInfo, Vec<TxnSummaryWithExpiration>)>,
+        batches_with_txn_summaries: Vec<(BatchInfo, Vec<TxnSummaryWithExpiration>)>,
     ) {
         let start = Instant::now();
-        for (batch_info, txn_summaries) in batch_summaries {
-            // As txn_summary_num_occurrences count the number of times the txn_summary occurred in the
-            // remaining proofs, we need to update the txn_summary_num_occurrences only if the batch is already
-            // in author_to_batches and batch_to_proof.
+
+        for (batch_info, txn_summaries) in batches_with_txn_summaries.into_iter() {
             let batch_sort_key = BatchSortKey::from_info(&batch_info);
-            let proof_exists_in_author_to_batches = self
-                .author_to_batches
-                .get(&batch_info.author())
-                .map_or(false, |batches| batches.contains_key(&batch_sort_key));
-            let proof_exists_in_batch_to_proof = self
-                .batch_to_proof
-                .get(&batch_sort_key.batch_key)
-                .map_or(false, Option::is_some);
-            let batch_summary_not_already_added = self
-                .batch_summaries
-                .get(&batch_sort_key.batch_key)
-                .map_or(true, |(_, _, added)| !*added);
-            if proof_exists_in_author_to_batches
-                && proof_exists_in_batch_to_proof
-                && batch_summary_not_already_added
-            {
+            self.author_to_batches
+                .entry(batch_info.author())
+                .or_default()
+                .insert(batch_sort_key.clone(), batch_info.clone());
+            self.expirations
+                .add_item(batch_sort_key, batch_info.expiration());
+
+            let mut txn_summary_exists = false;
+            let mut proof_exists = false;
+            let batch_key = BatchKey::from_info(&batch_info);
+            let item = self
+                .items
+                .entry(batch_key)
+                .and_modify(|item| {
+                    if item.txn_summaries.is_none() {
+                        item.txn_summaries = Some(txn_summaries.clone())
+                    } else {
+                        txn_summary_exists = true;
+                    }
+
+                    if item.proof.is_some() {
+                        proof_exists = true;
+                    }
+                })
+                .or_insert_with(|| QueueItem {
+                    info: batch_info,
+                    proof: None,
+                    proof_insertion_time: None,
+                    txn_summaries: Some(txn_summaries.clone()),
+                });
+
+            if proof_exists && txn_summary_exists {
```

Comment:
This doesn't seem accurate. 
The goal is to add `txn_summaries` to `txn_summary_num_occurrences` only once even the batches and proofs are received multiple times.
This logic seems to add `txn_summaries` to `txn_summary_num_occurrences` only if `txn_summaries` and the `proof` has already been received before. If the `txn_summaries` is received for the first time, we are not adding to `txn_summary_num_occurrences`.

Reply:
I see, makes sense.